kratos/utilities/mortar_utilities.h

```diff
@@ -592,22 +600,328 @@ class MortarUtilities
+        TVarType& ThisVariable,
+        const Matrix& ThisValue
+        );
+
 private:
 };// class MortarUtilities
 
 ///@name Explicit Specializations
 ///@{
 
+template<> 
+inline void MortarUtilities::ResetValue<Variable<double>, Historical>(
+        GeometryType& ThisGeometry,
+        Variable<double>& ThisVariable
+        )
+{
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); ++i_node)
+    {
+        ThisGeometry[i_node].FastGetSolutionStepValue(ThisVariable) = 0.0;
+    }
+}
+
+template<> 
+inline void MortarUtilities::ResetValue<component_type, Historical>(
+        GeometryType& ThisGeometry,
+        component_type& ThisVariable
+        )
+{
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); ++i_node)
+    {
+        ThisGeometry[i_node].FastGetSolutionStepValue(ThisVariable) = 0.0;
+    }
+}
+
+template<> 
+inline void MortarUtilities::ResetValue<Variable<array_1d<double, 3>>, Historical>(
+        GeometryType& ThisGeometry,
+        Variable<array_1d<double, 3>>& ThisVariable
+        )
+{
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); ++i_node)
+    {
+        auto& aux_value = ThisGeometry[i_node].FastGetSolutionStepValue(ThisVariable);
+        noalias(aux_value) = ZeroVector(3);
+    }
+}
+
+template<> 
+inline void MortarUtilities::ResetValue<Variable<double>, NonHistorical>(
+        GeometryType& ThisGeometry,
+        Variable<double>& ThisVariable
+        )
+{
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); ++i_node)
+    {
+        ThisGeometry[i_node].SetValue(ThisVariable, 0.0);
+    }
+}
+
+template<> 
+inline void MortarUtilities::ResetValue<component_type, NonHistorical>(
+        GeometryType& ThisGeometry,
+        component_type& ThisVariable
+        )
+{
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); ++i_node)
+    {
+        ThisGeometry[i_node].SetValue(ThisVariable, 0.0);
+    }
+}
+
+template<> 
+inline void MortarUtilities::ResetValue<Variable<array_1d<double, 3>>, NonHistorical>(
+        GeometryType& ThisGeometry,
+        Variable<array_1d<double, 3>>& ThisVariable
+        )
+{
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); ++i_node)
+    {
+        auto& aux_value = ThisGeometry[i_node].GetValue(ThisVariable);
+        noalias(aux_value) = ZeroVector(3);
+    }
+}
+
+template<> 
+inline void MortarUtilities::MatrixValue<Variable<double>, Historical>(
+        GeometryType& ThisGeometry,
+        Variable<double>& ThisVariable,
+        Matrix& ThisValue
+        )
+{
+    if (ThisValue.size1() != ThisGeometry.size() && ThisValue.size2() != 1)
+    {
+        ThisValue.resize(ThisGeometry.size(), 1, false);
+    }
+    for (unsigned int i_node = 0; i_node < ThisGeometry.size(); ++i_node)
+    {
+        ThisValue(i_node, 0) = ThisGeometry[i_node].FastGetSolutionStepValue(ThisVariable);
+    }
+}
+
+template<> 
+inline void MortarUtilities::MatrixValue<component_type, Historical>(
```

Comment:
why do u use a Matrix here instead of a vector

Reply:
Because that way can be used to multiply both matrix and vectors